allennlp/data/dataset_readers/nlvr.py

```diff
@@ -0,0 +1,142 @@
+                constant_type = nlvr_types.COMMON_TYPE_SIGNATURE[alias]
+                if constant_type != types.ANY_TYPE:
+                    self._agenda_mapping[constant] = "%s -> %s" % (constant_type, constant)
+
+    @overrides
+    def read(self, file_path):
+        instances = []
+        with open(file_path, "r") as data_file:
+            logger.info("Reading instances from lines in file: %s", file_path)
+            for line in tqdm.tqdm(data_file):
+                line = line.strip("\n")
+                if not line:
+                    continue
+                instances.append(self.text_to_instance(line))
+        if not instances:
+            raise ConfigurationError("No instances read!")
+        return Dataset(instances)
+
+    @overrides
+    def text_to_instance(self, json_line: str) -> Instance:
+        # pylint: disable=arguments-differ
+        data = json.loads(json_line)
+        sentence = data["sentence"]
+        label = data["label"] if "label" in data else None
+        structured_rep = data["structured_rep"]
+        world = NlvrWorld(structured_rep)
+        tokenized_sentence = self._tokenizer.tokenize(sentence)
+        sentence_field = TextField(tokenized_sentence, self._sentence_token_indexers)
+        agenda = self._get_agenda_for_sentence(sentence)
+        assert agenda, "No agenda found for sentence: %s" % sentence
+        is_nonterminal = lambda x: "<" not in x and len(x) > 1
+        production_rule_fields: List[Field] = []
+        instance_action_ids: Dict[str, int] = {}
+        for production_rule in world.all_possible_actions():
+            instance_action_ids[production_rule] = len(instance_action_ids)
+            field = ProductionRuleField(production_rule,
+                                        terminal_indexers=self._terminal_indexers,
+                                        nonterminal_indexers=self._nonterminal_indexers,
+                                        is_nonterminal=is_nonterminal,
+                                        context=tokenized_sentence)
+            production_rule_fields.append(field)
+        # agenda_field contains indices into actions.
+        agenda_field = ListField([LabelField(instance_action_ids[action],
+                                             skip_indexing=True,
+                                             label_namespace='actions')
+                                  for action in agenda])
+        action_field = ListField(production_rule_fields)
+        world_field = MetadataField(world)
+        fields = {"sentence": sentence_field,
+                  "agenda": agenda_field,
+                  "world": world_field,
+                  "actions": action_field}
+        if label:
+            label_field = LabelField(label, label_namespace='denotations')
+            fields["label"] = label_field
+        return Instance(fields)
+
+    def _get_agenda_for_sentence(self, sentence: str) -> List[str]:
```

Comment:
While coverage is related, that is not what I'm trying to do here. This method defines a mapping from sentence level features to (both non-terminal and terminal) productions. Moreover, not all words trigger actions, and some phrases trigger actions as well. I intend to make this mapping larger. I do agree that it would be nicer to learn the mapping as well. May be we can even start with these rules, and do some sort of a curriculum learning.